File: src/core/jupiter/core/domain/big_plans/big_plan_status.py
"""The status of a big plan."""
from functools import total_ordering

from jupiter.core.framework.value import EnumValue, enum_value


@enum_value
@total_ordering
class BigPlanStatus(EnumValue):
    """The status of a big plan."""

    # Created
    NOT_STARTED = "not-started"
    # Accepted
    ACCEPTED = "accepted"
    # Working
    IN_PROGRESS = "in-progress"
    BLOCKED = "blocked"
    # Completed
    NOT_DONE = "not-done"
    DONE = "done"

    @property
    def is_workable(self) -> bool:
        """Whether the status means the big plan is not completed."""
        return not self.is_completed

    @property
    def is_accepted(self) -> bool:
        """Whether the status means work has been accepted on the big plan."""
        return self == BigPlanStatus.ACCEPTED

    @property
    def is_accepted_or_more(self) -> bool:
        """Whether the status means work has been accepted, or is ongoing, or is completed."""
        return self.is_accepted or self.is_working or self.is_completed

    @property
    def is_working(self) -> bool:
        """Whether the status means work is ongoing for the big plan."""
        return self in (BigPlanStatus.IN_PROGRESS, BigPlanStatus.BLOCKED)

    @property
    def is_working_or_more(self) -> bool:
        """Whether the status means work is ongoing, or is completed."""
        return self.is_working or self.is_completed

    @property
    def is_completed(self) -> bool:
        """Whether the status means work is completed on the big plan."""
        return self in (BigPlanStatus.NOT_DONE, BigPlanStatus.DONE)

    def __lt__(self, other: object) -> bool:
        """Compare this with another."""
        if not isinstance(other, BigPlanStatus):
            raise Exception(
                f"Cannot compare an entity id with {other.__class__.__name__}",
            )

        all_values = self.get_all_values()

        return all_values.index(self.value) < all_values.index(other.value)

    @staticmethod
    def all_workable_statuses() -> list["BigPlanStatus"]:
        """All workable statuses."""
        return [s for s in BigPlanStatus if s.is_workable]

File: src/webui/app/logic/domain/big-plan-status.ts
import { BigPlanStatus } from "@jupiter/webapi-client";

export function bigPlanStatusName(status: BigPlanStatus): string {
  switch (status) {
    case BigPlanStatus.NOT_STARTED:
      return "Not Started";
    case BigPlanStatus.IN_PROGRESS:
      return "In Progress";
    case BigPlanStatus.BLOCKED:
      return "Blocked";
    case BigPlanStatus.NOT_DONE:
      return "Not Done";
    case BigPlanStatus.DONE:
      return "Done";
  }
}

export function bigPlanStatusIcon(status: BigPlanStatus): string {
  switch (status) {
    case BigPlanStatus.NOT_STARTED:
      return "🔧";
    case BigPlanStatus.IN_PROGRESS:
      return "🚧";
    case BigPlanStatus.BLOCKED:
      return "🚧";
    case BigPlanStatus.NOT_DONE:
      return "⛔";
    case BigPlanStatus.DONE:
      return "✅";
  }
}

const BIG_PLAN_STATUS_MAP = {
  [BigPlanStatus.NOT_STARTED]: 0,
  [BigPlanStatus.IN_PROGRESS]: 1,
  [BigPlanStatus.BLOCKED]: 2,
  [BigPlanStatus.NOT_DONE]: 3,
  [BigPlanStatus.DONE]: 4,
};

export function compareBigPlanStatus(
  status1: BigPlanStatus,
  status2: BigPlanStatus
): number {
  return BIG_PLAN_STATUS_MAP[status1] - BIG_PLAN_STATUS_MAP[status2];
}

export function isCompleted(status: BigPlanStatus): boolean {
  return status === BigPlanStatus.DONE || status === BigPlanStatus.NOT_DONE;
}
